# COGS 108 - Final Project: Vehicle Stops in San Diego

### Group Members: 
- Louise Xu (A12650425)
- David Tay (A10914210)
- Jeffery Lu

### Goal of Project:
- To determine whether there is a correlation between the frequency of vehicle stops in San Diego and the time and locations at which they take place
- To examine possible reasons for the correlations, if any exist 

### Research Question
- Are there more vehicle stops at certain times and places than others? Does the frequency of vehicle stops vary significantly between different seasons?

### Hypothesis
- There will most likely be more traffic violations during periods of time with high traffic density (e.g. rush hour), during weekdays (when there are more people on the road who are driving to work), and during holiday seasons (since there would be more tourists and people visting family). On a broader scale, given that the population of San Diego is increasing, it reasonable to say that the number of traffic violations is also increasing at a similar rate. 
- We can also guess that the number of vehicle stops in more populated areas is greater than those in less populated areas.

In [98]:
# Import packages
%matplotlib inline

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import patsy
import statsmodels.api as sm
from scipy.stats import ttest_ind

## Part 1: Load and Clean Data

In [99]:
# Import each vehicle stops data file into its own DataFrame
vs_14 = pd.read_csv('vehicle_stops_2014_datasd.csv')
vs_15 = pd.read_csv('vehicle_stops_2015_datasd.csv')
vs_16 = pd.read_csv('vehicle_stops_2016_datasd.csv')
vs_17 = pd.read_csv('vehicle_stops_2017_datasd.csv')

# Join the data for each year together into one DataFrame
# Note: The file for vs_17 contained two more columns (both of them empty and unnamed) than the datasets for the other years 
# These two columns were deleted on Microsoft Excel in order to successfully merge the four datasets together
vs_all = pd.concat([vs_14, vs_15, vs_16, vs_17], ignore_index=True)

#Import police regions into a DataFrame
sd_neigh = pd.read_csv('pd_beat_neighborhoods_datasd.csv')

In [100]:
# Let's check to see how many observations are in each dataset
# Note: vs_17 is year to date data from 2017, so there is data up to May of 2017 (the period in which this analysis is conducted)
print(vs_14.shape)
print(vs_15.shape)
print(vs_16.shape)
print(vs_17.shape)

(144164, 15)
(115422, 15)
(103051, 15)
(28362, 15)


We can see here that the number of vehicle stops is actually *decreasing* each year between 2014 and 2016. Also, while we can not take the 2017 data into direct consideration because we only have data up to May, we can predict the total number of vehicle stops at the end of this year based on our current data (28362 stops) divided by the five months that have already passed and then multiplied by the twelve months of the year, which results in (28362)(12)/5 = 68068.8 stops. We can round up to the nearest ten thousandsth and estimate that approximately 70000 vehicle stops will happen in the year of 2017. This still follows the trend of decreasing vehicle stops in San Diego.

In [101]:
# Let's start cleaning up the data

# We won't be using these columns
vs_all.drop(['arrested', 'searched', 'obtained_consent', 'contraband_found', 'property_seized'], axis=1, inplace=True, errors='ignore')

# Get rid of rows with invalid data
vs_all.dropna(inplace=True)
vs_all = vs_all.loc[(vs_all['service_area'] != 'Unknown')]

# Only keep data of stops involving San Diego residents
# vs_all = vs_all.loc[(vs_all['sd_resident'] == 'Y')]

# Cleaning the ages of the subjects, these characters occur in the dataset and we don't want these
vs_all = vs_all.loc[~(vs_all['subject_age'].str.contains("_|N|`|Q|F|Y|H", case=False))]
vs_all[['subject_age']] = vs_all[['subject_age']].apply(pd.to_numeric)

# Some stops have an X as the gender
vs_all = vs_all.loc[(vs_all['subject_sex'].str.contains("M|F", case=False))]

# Use Safe Harbor practices on the subject_ages
vs_all = vs_all.loc[(vs_all['subject_age'] < 90) | (vs_all['subject_age'] >= 16)] 

# Look at shape of data so far
vs_all


stop_id                  stop_cause service_area subject_race  \
0       1044975            Moving Violation          110            W   
1       1044976            Moving Violation          320            W   
2       1044977            Moving Violation          320            L   
3       1044978            Moving Violation          610            W   
4       1044980         Equipment Violation          930            H   
5       1044979         Equipment Violation          820            H   
6       1044981            Moving Violation          710            H   
7       1045139            Moving Violation          120            W   
8       1045141            Moving Violation          120            W   
9       1045140            Moving Violation          120            H   
10      1045142            Moving Violation          120            W   
11      1044982      Muni, County, H&S Code          710            H   
12      1044993         Equipment Violation          230            V   
13      1044984            Moving Violation          110            W   
14      1044983         Equipment Violation          240            W   
15      1044985            Moving Violation          720            B   
16      1044986            Moving Violation          230            W   
17      1190814         Equipment Violation          320            W   
19      1044988         Equipment Violation          820            B   
20      1044989         Equipment Violation          820            B   
21      1044990         Equipment Violation          430            H   
22      1044991            Moving Violation          230            W   
23      1044992         Equipment Violation          430            B   
24      1044996            Moving Violation          610            W   
25      1044994         Equipment Violation          230            W   
26      1044995            Moving Violation          610            W   
27      1044997            Moving Violation          240            W   
28      1045002            Moving Violation          720            H   
29      1044998            Moving Violation          320            H   
30      1045004            Moving Violation          720            H   
...         ...                         ...          ...          ...   
390954  1472913         Equipment Violation          510            B   
390956  1472915            Moving Violation          520            W   
390957  1472918            Moving Violation          510            H   
390958  1472930         Equipment Violation          930            W   
390959  1472912         Equipment Violation          440            B   
390961  1472916         Equipment Violation          520            H   
390963  1472925            Moving Violation          310            O   
390964  1472923            Moving Violation          320            H   
390965  1472924            Moving Violation          120            O   
390966  1472927            Moving Violation          120            O   
390968  1472993            Moving Violation          120            W   
390969  1472934            Moving Violation          310            O   
390970  1472929         Equipment Violation          310            W   
390971  1472931         Equipment Violation          440            H   
390972  1472946            Moving Violation          440            B   
390973  1472995            Moving Violation          120            W   
390974  1472926            Moving Violation          520            A   
390976  1472935            Moving Violation          310            B   
390977  1472941            Moving Violation          820            O   
390978  1472994            Moving Violation          120            W   
390980  1472938            Moving Violation          320            H   
390983  1472942            Moving Violation          320            B   
390984  1472944            Moving Violation          320          

## Part 2: Exploratory Data Vizualization

## Part 3: Exploring The Data

## Part 4: Data Analysis

## Part 5: Discussion & Conclusions